README.md
[![Build Status](https://travis-ci.org/neechbear/blip.svg?branch=master)](https://travis-ci.org/neechbear/blip)
[![Code Climate](https://codeclimate.com/github/neechbear/blip/badges/gpa.svg)](https://codeclimate.com/github/neechbear/blip)
[![Issue Count](https://codeclimate.com/github/neechbear/blip/badges/issue_count.svg)](https://codeclimate.com/github/neechbear/blip)
[![Pre-release Alpha](https://img.shields.io/badge/status-pre--release%20alpha-ff69b4.svg)](https://nicolaw.uk/blip)

# blip - Bash Library for Indolent Programmers

Programmers are lazy. Good system administrators are _really_ lazy. (Why bother
doing something more than once)?

Unfortunately, due to the fact that Bash doesn't particularly lend itself to
reusable code, it doesn't enjoy the same wealth of shared code available that
you find with Python or Perl.

_"But what about the lazy sysadmin that needs to write a script, where Bash
genuinely is the most appropriate option?"_, I hear you ask! Well, by providing
functions for many common tasks, I'm hoping that `blip` will help fill some of
the gaps for those situations.

```
source /usr/lib/blip.bash
```

Please see the man page `man blip.bash`, [bash.pod for full
documentation](blip.bash.pod) or `/usr/share/doc/blip` directory for code
examples and other useful information.

* <https://nicolaw.uk/blip>
* <https://github.com/neechbear/blip/>
  * <https://github.com/neechbear/blip/releases>
  * <https://launchpad.net/~nicolaw/+archive/ubuntu/blip>
  * <http://ppa.launchpad.net/nicolaw/blip/ubuntu/>
  * <https://raw.githubusercontent.com/neechbear/blip/master/blip.bash>

## Installation

### Ubuntu

On Ubuntu, you can install from my PPA
[ppa:nicolaw/blip](https://launchpad.net/~nicolaw/+archive/ubuntu/blip)
by running the following commands:

```
sudo add-apt-repository ppa:nicolaw/blip
sudo apt-get update
sudo apt-get install blip
```

### Debian

On Debian or other Debian-based distributions, you can
[download the DEB package from GitHub](https://github.com/neechbear/blip/releases)
and install it manually with:

```
curl -o https://github.com/neechbear/blip/releases/download/path/to/blip_pkg.deb
sudo dpkg -i blip_pkg.deb
```

### RedHat, CentOS, Fedora

Similarly for RedHat based distributions you can
[install the RPM package from GitHub](https://github.com/neechbear/blip/releases)
manually with:

```
sudo yum localinstall https://github.com/neechbear/blip/releases/download/path/to/blip_pkg.rpm
```

### Source

_Write instructions for installation from GitHub or source tarball here._

## TODO

* Merge all the other cool and reusable stuff I've written in to this library
  (see pending functionality below).
* Make all the shell scripting comply with a sensible style guide (like Google's
  one at <https://google.github.io/styleguide/shell.xml>).
* Finish writing comprehensive manual page with code examples for each function.
* Finish writing comprehensive unit tests with full code coverage.
* Setup automatic build of release tarballs, Deb and RPM packages upon GitHub
  repository commits (assuming a Travis CI pass of unit tests).
* Maybe create [pkg](https://wiki.archlinux.org/index.php/creating_packages),
  [apk](https://wiki.alpinelinux.org/wiki/Creating_an_Alpine_package) and
  [portage](https://wiki.gentoo.org/wiki/Portage) packages if I get bored.

### Pending Functionality

* Add `get_user_input()` - multi character user input without defaults.
* Add process locking functions.
* Add background daemonisation functions (ewww; ppl should use systemd).
* Add standard logging functions.
* Add syslogging functionality of all process `STDOUT` + `STDERR`.
* Add common array manipulation functions.

## See Also

<https://github.com/akesterson/cmdarg> - A pure bash library to make argument
parsing far less troublesome.



debian/changelog
blip (0.4-1) trusty; urgency=low

  * Build script verbosity made configurable.
  * Build script tages version and release information from git repo tag.
  * Added json string and object output functions.
  * Bug fix in url_exists with file:// shemes.

 -- Nicola Worthington <nicolaw@tfb.net>  Tue, 15 Nov 2016 18:19:58 +0100

blip (0.3-1) trusty; urgency=low

  * Fixed build dependencies.
  * Skip network based unit tests when curl is unable to probe test URL.

 -- Nicola Worthington <nicolaw@tfb.net>  Tue, 25 Oct 2016 21:28:45 +0100

blip (0.2-1) trusty; urgency=low

  * Build script changes for LaunchPad packaging guidelines.
  * Added is_newer_version() function.
  * Added required_command_version() function.

 -- Nicola Worthington <nicolaw@tfb.net>  Tue, 25 Oct 2016 20:10:17 +0100

blip (0.1-4~ubuntu1) trusty; urgency=low

  * Fixed package dependencies to more accurately reflect BLIP_EXTERNAL_CMD*
    requirements.

 -- Nicola Worthington <nicolaw@tfb.net>  Tue, 25 Oct 2016 09:45:55 +0100

blip (0.1-4) trusty; urgency=low

  * Additional documentation.
  * Added many unit tests.
  * Added / moved some example scripts.
  * Added MAC and additional variable type tests.
  * Build changes for deb signing.

 --  Nicola Worthington <nicolaw@tfb.net>  Tue, 25 Oct 2016 00:41:00 +0100

blip (0.1-3) UNRELEASED; urgency=low

  * Added BLIP_VERSION and BLIP_VERSINFO variables.
  * Added ANSI terminal colour code variables.
  * Added license agreement to header comments.
  * Added trap stack handlers.
  * Added /usr/share/doc/blip examples.
  * Other minor changes.

 --  Nicola Worthington <nicolaw@tfb.net>  Tue, 10 Oct 2016 22:45:00 +0100

blip (0.01-2) UNRELEASED; urgency=low

  * Testing build.sh packaging signing functionality.

 --  Nicola Worthington <nicolaw@tfb.net>  Wed, 28 Sep 2016 20:31:00 +0100

blip (0.01-1) UNRELEASED; urgency=low

  * Testing build.sh and GitHub.com release functionality.

 --  Nicola Worthington <nicolaw@tfb.net>  Tue, 27 Sep 2016 23:08:00 +0100

blip (0.01) UNRELEASED; urgency=low

  * Initial development pre-release.

 -- Nicola Worthington <nicolaw@tfb.net>  Sun, 25 Sep 2016 18:43:28 +0100


test.sh
#!/bin/bash

# This is just pratting about local testing until I put together a proper suite
# of unit tests and what not.

set -euo pipefail

BLIP_ANSI_VARIABLES=1
BLIP_REQUIRE_VERSION="0.01-3"
source blip.bash
BLIP_DEBUG_LOGLEVEL=3

compgen -A variable | grep ANSI
echo "${!ANSI[@]}"
echo "${ANSI_FG_YELLOW}${ANSI_BOLD_ON}Hello world.${ANSI_RESET}"
echo "${ANSI[blink]}${ANSI[bg_white]}${ANSI[bold]}${ANSI[red]}${ANSI[underline]}Hello world.${ANSI[reset]}"

push_trap_handler "echo 'hello world'" INT
push_trap_handler "echo 'foo bar'" INT HUP
push_trap_handler "echo 'this will not appear'" INT HUP
pop_trap_handler INT
get_trap_handler INT
set_trap_handler "echo 'woop woop'" INT
get_trap_handler INT
get_trap_handler HUP
unset_trap_handler HUP
get_trap_handler HUP
push_trap_handler 'for ((x=0; x<=10; x++)) ; do echo " >> x=$x << "; done' INT HUP
push_trap_handler "echo 'The final countdown.'" INT

echo "${ANSI[bold]}${ANSI[cyan]}Try pressing Control-C to trigger a SIGINT trap handler stack.${ANSI[reset]}"
for ((i=0; i<=10; i++)); do
    echo "$i"
    sleep 1
done
echo "${ANSI[bold]}${ANSI[red]}Too late; better luck next time!${ANSI[reset]}"

rc_to_colour () {
    if [[ ${1:-} -eq 0 ]] ; then
        echo '[0;1;32m'
    else
        echo '[0;1;31m'
    fi
}

while read -r command ; do
    result="$(eval "$command")"
    rc=$?
    echo -e "$(rc_to_colour "$rc")$command\e[0m [$rc] $result"
done << 'COMMANDS'
get_iso8601_date
get_unixtime
get_date
get_free_disk_space /
get_gecos_info email
get_gecos_name
get_gecos_name postgres
COMMANDS

set +e

while read -r addr ; do
    while read -r command ; do
        $command "$addr"
        rc=$?
        echo -e "$(rc_to_colour "$rc")$command \"$addr\"\e[0m [$rc]"
    done << 'COMMANDS'
is_ipv4_address
is_ipv4_prefix
is_ipv6_address
is_ipv6_prefix
COMMANDS
done << 'ADDRESSES'
10.10.10.10
10.10.10.10/8
10.10.10.10/40
300.300.300.300
::1
fe80::a/64
2000::aaaa::ffff
ADDRESSES

while read -r command ; do
    result="$(eval "$command")"
    rc=$?
    echo -e "$(rc_to_colour "$rc")$command\e[0m [$rc] $result"
done << 'COMMANDS'
url_exists http://www.bbc.co.uk
url_http_response_code http://www.bbc.co.uk
url_http_header http://www.bbc.co.uk
COMMANDS

get_user_selection "Abort, retry, fail?" "f" "a" "r" "f"
echo rc=$?

get_user_selection "Abort, retry, fail?" "" "a" "r" "f"
echo rc=$?

get_user_confirmation "Continue?"
echo rc=$?

get_user_confirmation "Continue?" "y"
echo rc=$?

get_user_confirmation "Continue?" "n"
echo rc=$?



tests/test.sh
#!/bin/bash

# This is just pratting about local testing until I put together a proper suite
# of unit tests and what not.

set -euo pipefail

BLIP_ANSI_VARIABLES=1
BLIP_REQUIRE_VERSION="0.01-3"
source blip.bash
BLIP_DEBUG_LOGLEVEL=3

compgen -A variable | grep ANSI
echo "${!ANSI[@]}"
echo "${ANSI_FG_YELLOW}${ANSI_BOLD_ON}Hello world.${ANSI_RESET}"
echo "${ANSI[blink]}${ANSI[bg_white]}${ANSI[bold]}${ANSI[red]}${ANSI[underline]}Hello world.${ANSI[reset]}"

push_trap_stack "echo 'hello world'" INT
push_trap_stack "echo 'foo bar'" INT HUP
push_trap_stack "echo 'this will not appear'" INT HUP
pop_trap_stack INT
get_trap_stack INT
set_trap_stack "echo 'woop woop'" INT
get_trap_stack INT
get_trap_stack HUP
unset_trap_stack HUP
get_trap_stack HUP
push_trap_stack 'for ((x=0; x<=10; x++)) ; do echo " >> x=$x << "; done' INT HUP
push_trap_stack "echo 'The final countdown.'" INT

echo "${ANSI[bold]}${ANSI[cyan]}Try pressing Control-C to trigger a SIGINT trap handler stack.${ANSI[reset]}"
for ((i=6; i>=0; i--)); do
    echo "$i"
    sleep 1
done
echo "${ANSI[bold]}${ANSI[red]}Too late; better luck next time!${ANSI[reset]}"

set_trap_stack "echo 'Interrupted by user input.'" INT
set_trap_stack "echo 'Interrupted by signal.'" HUP

rc_to_colour () {
    if [[ ${1:-} -eq 0 ]] ; then
        echo '[0;1;32m'
    else
        echo '[0;1;31m'
    fi
}

while read -r command ; do
    result="$(eval "$command")"
    rc=$?
    echo -e "$(rc_to_colour "$rc")$command\e[0m [$rc] $result"
done << 'COMMANDS'
get_iso8601_date
get_unixtime
get_date
get_free_disk_space /
get_gecos_info email
get_gecos_name
get_gecos_name postgres
COMMANDS

set +e

while read -r addr ; do
    while read -r command ; do
        $command "$addr"
        rc=$?
        echo -e "$(rc_to_colour "$rc")$command \"$addr\"\e[0m [$rc]"
    done << 'COMMANDS'
is_ipv4_address
is_ipv4_prefix
is_ipv6_address
is_ipv6_prefix
COMMANDS
done << 'ADDRESSES'
10.10.10.10
10.10.10.10/8
10.10.10.10/40
300.300.300.300
::1
fe80::a/64
2000::aaaa::ffff
ADDRESSES

while read -r command ; do
    result="$(eval "$command")"
    rc=$?
    echo -e "$(rc_to_colour "$rc")$command\e[0m [$rc] $result"
done << 'COMMANDS'
url_exists http://www.bbc.co.uk
url_http_response_code http://www.bbc.co.uk
url_http_header http://www.bbc.co.uk
COMMANDS

get_user_selection "Abort, retry, fail?" "f" "a" "r" "f"
echo rc=$?

get_user_selection "Abort, retry, fail?" "" "a" "r" "f"
echo rc=$?

get_user_confirmation "Continue?"
echo rc=$?

get_user_confirmation "Continue?" "y"
echo rc=$?

get_user_confirmation "Continue?" "n"
echo rc=$?



tests/child.sh
#!/bin/bash

source blip.bash

counter () {
  declare -i i
  for ((i=0; i<=4; i++)) ; do 
    echo "$(( $(printf '%(%s)T' -1) - $(printf '%(%s)T' -2) ))"
    sleep 0.6
  done
}

echo "Hello world!"

counter "$@"

declare -x foo="BAR MEH BLERG"
echo "Woop ${foo##* }!"



build.sh
#!/bin/bash
#
# MIT License
#
# Copyright (c) 2016 Nicola Worthington
#

set -ueo pipefail
declare -g verbose
if [[ $- =~ [vx] ]] ; then verbose=1 ; fi
shopt -s checkwinsize
umask 0077

_build_deb_packages () {
  if ! type -P debuild >/dev/null 2>&1 ; then
    return 0
  fi

  _debuild () {
    declare release_dir="$1"
    declare debuild_extra_args="${2:-}"
    mkdir -p ${verbose:+-v} "$release_dir"

    declare debian_orig_tar="$build_base/${tarball%.tar.gz}.orig.tar.gz"
    debian_orig_tar="${debian_orig_tar//-/_}"
    cp ${verbose:+-v} "$build_base/$tarball" "$debian_orig_tar"

    pushd "$build_dir"
    debuild -sa ${debuild_extra_args:- -us -uc}
    popd

    mv ${verbose:+-v} -- \
      "$build_base"/*.{dsc,changes,build,debian.tar.gz,orig.tar.gz} \
      "$release_dir"

    if stat -t "$build_base"/*.deb >/dev/null 2>&1 ; then
      mv ${verbose:+-v} -- "$build_base"/*.deb "$release_dir"
      dpkg-deb -I "$release_dir/${pkg}_${VERSION}${DIRTY_SUFFIX:+$DIRTY_SUFFIX}_all.deb"
      dpkg-deb -c "$release_dir/${pkg}_${VERSION}${DIRTY_SUFFIX:+$DIRTY_SUFFIX}_all.deb"
    fi
  }

  _debuild "$release_dir"
  if [[ -n "${gpg_keyid:-}" ]] ; then
    _debuild "$release_dir/deb-src" "-k${gpg_keyid} -S"
  fi
}

_build_rpm_packages () {
  if ! type -P rpmbuild >/dev/null 2>&1 ; then
    return 0
  fi

  if [[ -n "${gpg_keyid:-}" && -n "${gpg_name:-}" ]] ; then
    while read -r line ; do
      if ! grep -q "^$line$" ~/.rpmmacros ; then
        echo "$line" >> ~/.rpmmacros
      fi
    done <<RPMMACROS
%_signature gpg
%_gpg_name $gpg_name
%_gpgbin /usr/bin/gpg
RPMMACROS
  fi

  rpmbuild -ba "$build_dir/${pkg}.spec" \
    ${gpg_keyid:+--sign} \
    --define "name $pkg" \
    --define "version $VERSION_MAJOR.$VERSION_MINOR${VERSION_POINT:+.$VERSION_POINT}" \
    --define "release $VERSION_RELEASE" \
    --define "_sourcedir $build_base" \
    --define "_rpmdir $release_dir" \
    --define "_build_name_fmt %%{NAME}-%%{VERSION}-%%{RELEASE}${DIRTY_SUFFIX:+$DIRTY_SUFFIX}.%%{ARCH}.rpm"

  rpm -qlpi ${verbose:+-v} \
    "$release_dir/${pkg}-${VERSION}${DIRTY_SUFFIX:+$DIRTY_SUFFIX}.noarch.rpm"
}

_build () {
  # Copy files in to build directory.
  rsync -a ${verbose:+-v} \
    --exclude=".*" --exclude="build/" --exclude="release/" \
    --exclude="gitversion.sh" --exclude="build.sh" \
    "${base%/}/" "${build_dir%}/"

  # Macro substitution of version information.
  sed -i -e "s/@VERSION_MAJOR@/$VERSION_MAJOR/g" "${build_dir%}/${pkg}.bash.in"
  sed -i -e "s/@VERSION_MINOR@/$VERSION_MINOR/g" "${build_dir%}/${pkg}.bash.in"
  sed -i -e "s/@VERSION_RELEASE@/$VERSION_RELEASE/g" "${build_dir%}/${pkg}.bash.in"
  sed -i -e "s/@VERSION_TAG@/$TAG_SHA1_SHORT${DIRTY_SUFFIX:+$DIRTY_SUFFIX}/g" "${build_dir%}/${pkg}.bash.in"
  mv ${verbose:+-v} "${build_dir%}/${pkg}.bash.in" "${build_dir%}/${pkg}.bash"

  # Dynamically update RPM and DEB package changelogs.
  "$base/gitversion.sh" -d "$base/.git" -p "$pkg" -S -l "deb" \
    > "${build_dir%}/debian/changelog"
  "$base/gitversion.sh" -d "$base/.git" -p "$pkg" -S -l "rpm" \
    >> "${build_dir%}/${pkg}.spec.in"
  mv ${verbose:+-v} "${build_dir%}/${pkg}.spec.in" "${build_dir%}/${pkg}.spec"

  # Generate some Groff man pages from the POD source.
  pod2man \
    --name="BLIP.BASH" \
    --release="${pkg}.bash $VERSION" \
    --center="${pkg}.bash" \
    --section=3 \
    --utf8 "$build_dir/${pkg}.bash.pod" > "$build_dir/${pkg}.bash.3"

  # Generate an HTML version of the README markdown.
  if type -P markdown >/dev/null 2>&1 ; then
    markdown "$build_dir/README.md" > "$build_dir/README.html"
  fi

  # Create release tarball of resulting build directory.
  tar -C "$build_base" ${verbose:+-v} -zcf "$build_base/$tarball" "${build_dir##*/}/"
  cp ${verbose:+-v} "$build_base/$tarball" "$release_dir"
}

main () {
  declare base
  base="$(readlink -f "${BASH_SOURCE[0]%/*}")"

  # Gather information and declare variables first.
  declare -g pkg="blip"
  eval "$("$base/gitversion.sh" -d "$base/.git" -p "$pkg" -S)"
  declare -g build_base="$base/build"
  declare -g build_dir="$build_base/${pkg}-${VERSION_MAJOR}.${VERSION_MINOR}${VERSION_POINT:+$VERSION_POINT}"
  declare -g release_dir="$base/release/${pkg}-${VERSION}${VERSION_POINT:+$VERSION_POINT}"
  declare -g tarball="${pkg}-${VERSION_MAJOR}.${VERSION_MINOR}.tar.gz"

  if [[ "$(hostid)" = "007f0101" ]] && [[ "$(id -un)" = "nicolaw" ]] ; then
    # Sign with the authors key.
    declare -g gpg_keyid="6393F646"
    declare -g gpg_name="Nicola Worthington"
  fi

  # Clean, then build the package.
  rm -Rf ${verbose:+-v} --one-file-system --preserve-root "$build_base"
  mkdir -p ${verbose:+-v} "$build_base" "$build_dir" "$release_dir"
  _build
  _build_deb_packages
  _build_rpm_packages

  # List the resulting release package files.
  ls --color -Rla "$release_dir"
}

main "$@"



gitversion.sh
#!/bin/bash
#
# MIT License
#
# Copyright (c) 2016 Nicola Worthington
#

set -ueo pipefail
shopt -s nocasematch

git () {
  command git --git-dir="${GIT_DIR:-.git}" "$@"
}

parse_args () {
  # TODO(nicolaw): This is fragile; maybe improve with getopt/getops?
  while [[ $# -ge 1 ]] ; do
    declare key="${1:-}"
    case "$key" in
      -d|--git-dir)   printf 'GIT_DIR="%s"\n' "${2:-.git}"; shift ;;
      -p|--pkg-name)  printf 'PKG_NAME="%s"\n' "${2:-}"; shift ;;
      -k|--key)       printf 'PRINT_KEY="%s"\n' "${2:-}"; shift ;;
      -S|--no-source) printf 'NO_SOURCE="1"\n' ;;
      -b|--branch)    printf 'BRANCH="%s"\n' "${2:-master}"; shift ;;
      -c|--commit)    printf 'COMMIT="%s"\n' "${2:-HEAD}"; shift ;;
      -t|--tag)       printf 'TAG="%s"\n' "${2:-}"; shift ;;
      -l|--log)       printf 'LOG="%s"\n' "${2:-changelog}"; shift ;;
      -O|--os-name)   printf 'OS_NAME="%s"\n' "${2:-}"; shift ;;
      -D|--os-dist)   printf 'OS_DISTRIBUTION="%s"\n' "${2:-}"; shift ;;
      -V|--os-ver)    printf 'OS_VERSION="%s"\n' "${2:-}"; shift ;;
      -v|--version)   printf 'PRINT_VERSION="!"\n' ;;
      -h|--help)      printf 'PRINT_HELP="1"\n' ;;
    esac
    shift
  done
}

print_version () {
  echo "https://media.giphy.com/media/101qkTeAqyMbio/giphy.gif"
}

print_usage () {
  cat <<EOM
Syntax: ${0#*/} [options]
  -d, --git-dir=DIR    Specify .git repository (defaults to .git)
  -p, --pkg-name=NAME  Specify the package name (derive from repo path/origin)
  -k, --key=KEY        Only print output value of specific KEY
  -S, --no-source      Do not attempt to discover upstream source
  -b, --branch=BRANCH  Specify branch (defaults to master)
  -c, --commit=ID      Specify last git commit-ish like ID to work up to
  -t, --tag=TAG        Specify git tag to work from
  -l, --log=FORMAT     Output changelog in a particular format (deb, rpm)
  -O, --os-name=NAME   Overload OS detection, explicitly providing OS_NAME
  -D, --os-dist=DIST   Overload OS detection, explicitly providing OS_DISTRIB
  -V, --os-ver=VERSION Overload OS detection, explicitly providing OS_VERSION
  -h, --help           Display this help message
EOM
}

os_name () {
  uname -s
}

os_distribution () {
  local os_name="${1:-}"
  if [[ -r /etc/lsb-release ]] ; then
    eval "$(egrep "^DISTRIB_ID=.*" "/etc/lsb-release")"
  fi
  if [[ -n "$DISTRIB_ID" ]] ; then
    echo "$DISTRIB_ID"
  elif [[ -e /etc/centos-release ]] ; then
    echo "CentOS"
  elif [[ -e /etc/redhat-release ]] ; then
    echo "RHEL"
  elif [[ -e /etc/ubuntu-release ]] ; then
    echo "Ubuntu"
  elif [[ -e /etc/debian-release ]] ; then
    echo "Debian"
  fi
  if [[ "$os_name" == "Darwin" ]] ; then
    sw_vers -productName
  fi
}

os_codename () {
  local os_name="${1:-}"
  if [[ -r /etc/lsb-release ]] ; then
    eval "$(egrep "^DISTRIB_CODENAME=.*" "/etc/lsb-release")"
  fi
  if [[ -n "$DISTRIB_CODENAME" ]] ; then
    echo "$DISTRIB_CODENAME"
  fi
}

os_version () {
  local os_name="${1:-}"
  if [[ -r /etc/lsb-release ]] ; then
    eval "$(egrep "^DISTRIB_RELEASE=.*" "/etc/lsb-release")"
  fi
  if [[ -n "$DISTRIB_RELEASE" ]] ; then
    echo "$DISTRIB_RELEASE"
  elif [[ "$os_name" == "Darwin" ]] ; then
    sw_vers -productVersion
  else
    uname -r
  fi
}

os_pkg_suffix () {
  local os="${1:-}"
  local ver="${2:-}"
  case "$os" in
    RedHat*|RHEL|CentOS*) echo "el${ver%%.*}" ;;
    Fedora*) echo  "fc${ver%%.*}" ;;
    Ubuntu*) echo "ubuntu" ;;
  esac
}

git_branch () {
  git branch | grep '^\*\s*.*' | cut -d ' ' -f 2-
}

git_tag () {
  local commit="${1:-$COMMIT}"
  git describe --tags --abbrev=0 --first-parent --match=v* "${commit:-HEAD}"
}

git_tag_bash_rematch () {
  [[ "${1:-}" =~ ^v(([0-9]+)\.([0-9]+)(\.([0-9]+))?(-([0-9]+))?)$ ]]
}

git_longid () {
  local commit="${1:-$COMMIT}"
  git rev-list -n 1 "${commit:-HEAD}"
}

git_shortid () {
  local commit="${1:-$COMMIT}"
  git rev-parse --short "${commit:-HEAD}"
}

git_isdirty () {
  [[ "$(git diff --shortstat 2> /dev/null | tail -n1)" != "" ]] && echo "true"
}

git_source () {
  git remote show origin 2>/dev/null \
    | grep "Fetch URL" | cut -d : -f 2- | cut -d ' ' -f 2-
}

changelog () {
  local logtype="${1:-}"
  local first="${2:-}"
  local last="${3:-}"
  local annotation="${4:-}"

  if [[ -z "$first" && -z "$last" ]] ; then
    local output=""
    while read -r last annotation; do
      output="$(changelog "$logtype" "$first" "$last" "$annotation")

$output"
      first="$last"
    done < <(
      git tag -l -n1 v*
      if [[ "$(git_longid "$(git_tag)")" != "$(git_longid "HEAD")" ]] ; then
        echo "HEAD"
      fi
    )
    echo "$output"

  else
    local tag="$last"
    if [[ "$tag" == "HEAD" ]] ; then
      tag="$first"
    fi
    if [[ "$tag" == "HEAD" || "$tag" == "$TAG" ]] && [[ -n "$DIRTY" ]] ; then
      local dirty_suffix="${DIRTY_SUFFIX:-}"
    fi

    if git_tag_bash_rematch "$tag" ; then
      local version="${BASH_REMATCH[1]}"
      local version_major="${BASH_REMATCH[2]}"
      local version_minor="${BASH_REMATCH[3]}"
      local version_point="${BASH_REMATCH[5]}"
      local version_release="${BASH_REMATCH[7]}"
    else
      >&2 echo "Unable to extract version information for git tag $tag; aborting!"
      exit 99
    fi

    case "${logtype:-}" in
      debian|ubuntu|dch|deb)
        local urgency="low"
        local os_distribution="${OS_CODENAME:-}"
        os_distribution="${os_distribution,,}"

        if [[ "$annotation" =~ urgency=([^[:space:]]+) ]] ; then
          urgency="${BASH_REMATCH[1],,}"
        fi
        if [[ "$last" == "HEAD" ]] ; then
          os_distribution="UNRELEASED"
          version="$version"
        fi

        printf '%s (%s) %s; urgency=%s\n\n' \
          "$PKG_NAME" "$version${dirty_suffix:-}" "$os_distribution" "$urgency"
        git log --format="format:  * %s" "${first:+$first..}${last}" | cat
        git log -1 \
          --format="format:%n%n -- %aN <%aE>  %aD" \
          "${last}" | cat
        ;;

      redhat|rhel|centos|fedora|rpm)
        git log -1 \
          --format="format:* @%ad@ %aN <%aE> - $version${dirty_suffix:-}%n" \
          --date=raw \
          "${last}" \
            | perl -MPOSIX=strftime -pe \
            's/^\* @([0-9]+) [\+-]?[0-9]{4}@ /* @{[strftime("%a %b %d %Y",localtime($1))]} /'
          # Custom date formats are not supported in Git 1.x :-(
          #--date=format:"%a %b %d %Y" \
        git log \
          --format="format:- %s" \
          "${first:+$first..}${last}" | cat
        ;;

      *%[a-zA-Z]*)
        git log \
          ${logtype:+--format="format:$logtype"} \
          "${first:+$first..}${last}" | cat
        ;;
    esac
  fi
}

main () {
  eval "$(parse_args "$@")"
  if [[ $? -gt 0 || -n "${PRINT_HELP:-}" ]] ; then
    print_usage
    return 0
  elif [[ -n "${PRINT_VERSION:-}" ]] ; then
    print_version
    return 64
  fi

  GIT_DIR="${GIT_DIR:-.git}"
  GIT_TOPLEVEL="$(git rev-parse --show-toplevel)" 
  BRANCH="${BRANCH:-$(git_branch)}"

  if [[ -z "${NO_SOURCE:-}" ]] ; then
    SOURCE="$(git_source)"
  fi

  if [[ -z "${PKG_NAME:-}" && "${SOURCE:-}" =~ /([^/]+)\.git$ ]] ; then
    PKG_NAME="${BASH_REMATCH[1]}"
  elif [[ -z "${PKG_NAME:-}" && -n "${GIT_TOPLEVEL:-}" ]] ; then
    PKG_NAME="${GIT_TOPLEVEL##*/}"
  fi

  COMMIT="${COMMIT:-HEAD}"
  COMMIT_SHA1="$(git_longid "$COMMIT")"
  COMMIT_SHA1_SHORT="$(git_shortid "$COMMIT")"

  TAG="$(git_tag "$COMMIT")"
  TAG_SHA1="$(git_longid "$TAG")"
  TAG_SHA1_SHORT="$(git_shortid "$TAG")"

  TAG_CHANGES=$(git rev-list "$TAG..$COMMIT" --count)
  DIRTY="$(git_isdirty)" || :
  if [[ -n "$DIRTY" ]] ; then
    DIRTY_SUFFIX="~devbuild${TAG_CHANGES:-0}"
  fi

  if git_tag_bash_rematch "$TAG" ; then
    VERSION="${BASH_REMATCH[1]}"
    VERSION_MAJOR="${BASH_REMATCH[2]}"
    VERSION_MINOR="${BASH_REMATCH[3]}"
    VERSION_POINT="${BASH_REMATCH[5]}"
    VERSION_RELEASE="${BASH_REMATCH[7]}"
  fi

  # TODO(nicolaw): What about arch etc? This feels flakey.
  if [[ -z "${OS_DISTRIBUTION:-}" && -z "${OS_NAME:-}" && -z "${OS_VERSION:-}" ]] ; then
    OS_NAME="$(os_name)"
    OS_DISTRIBUTION="$(os_distribution "${OS_NAME:-}")"
    OS_CODENAME="$(os_codename "${OS_NAME:-}")"
    OS_VERSION="$(os_version "${OS_NAME:-}")"
  fi
  if [[ -n "${OS_NAME:-}" && -z "${OS_DISTRIBUTION:-}" ]] ; then
    OS_DISTRIBUTION="$OS_NAME"
  fi
  OS_PKG_SUFFIX="$(os_pkg_suffix "${OS_DISTRIBUTION:-}" "${OS_VERSION:-}")"

  if [[ -n "${PRINT_KEY:-}" ]] ; then
    echo "${!PRINT_KEY:-}"

  elif [[ -n "${LOG:-}" ]] ; then
    changelog "$LOG"

  else
    for v in GIT_DIR GIT_TOPLEVEL PKG_NAME DIRTY DIRTY_SUFFIX SOURCE BRANCH \
      COMMIT COMMIT_SHA1 COMMIT_SHA1_SHORT \
      TAG TAG_SHA1 TAG_SHA1_SHORT TAG_CHANGES \
      VERSION VERSION_MAJOR VERSION_MINOR VERSION_POINT VERSION_RELEASE \
      OS_NAME OS_DISTRIBUTION OS_CODENAME OS_VERSION OS_PKG_SUFFIX
    do
      if [[ -n "${!v+defined}" ]] ; then
        printf '%s="%s"\n' "$v" "${!v:-}"
      fi
    done
  fi
}

main "$@"

# git describe --tags --dirty --always 2> /dev/null || echo "dev"
# git tag -u 6393F646 -a v0.3-1 -m 'Initial successful LaunchPad PPA submission' 7947d41
# git config --global user.signingkey "6393F646"
# git --no-pager log v0.3-1..HEAD --pretty --format='%cD,%cn,%ce,%h,"%s","%d"'
# git --no-pager log  --pretty --format='%cD,%cn,%ce,%h,"%s","%d"'
# git tag -l -n9
# git tag -l 




examples/example2.sh
#!/bin/bash

# This is just pratting about local testing until I put together a proper suite
# of unit tests and what not.

set -euo pipefail

BLIP_ANSI_VARIABLES=1
BLIP_REQUIRE_VERSION="0.01-3"
source "${BASH_SOURCE[0]%/*}/../blip.bash"
BLIP_DEBUG_LOGLEVEL=3

compgen -A variable | grep ANSI
echo "${!ANSI[@]}"
echo "${ANSI_FG_YELLOW}${ANSI_BOLD_ON}Hello world.${ANSI_RESET}"
echo "${ANSI[blink]}${ANSI[bg_white]}${ANSI[bold]}${ANSI[red]}${ANSI[underline]}Hello world.${ANSI[reset]}"

rc_to_colour () {
    if [[ ${1:-} -eq 0 ]] ; then
        echo '[0;1;32m'
    else
        echo '[0;1;31m'
    fi
}

while read -r command ; do
    result="$(eval "$command")"
    rc=$?
    echo -e "$(rc_to_colour "$rc")$command\e[0m [$rc] $result"
done << 'COMMANDS'
get_iso8601_date
get_unixtime
get_date
get_free_disk_space /
get_gecos_info email
get_gecos_name
get_gecos_name postgres
COMMANDS

set +e

while read -r addr ; do
    while read -r command ; do
        $command "$addr"
        rc=$?
        echo -e "$(rc_to_colour "$rc")$command \"$addr\"\e[0m [$rc]"
    done << 'COMMANDS'
is_ipv4_address
is_ipv4_prefix
is_ipv6_address
is_ipv6_prefix
COMMANDS
done << 'ADDRESSES'
10.10.10.10
10.10.10.10/8
10.10.10.10/40
300.300.300.300
::1
fe80::a/64
2000::aaaa::ffff
ADDRESSES

while read -r command ; do
    result="$(eval "$command")"
    rc=$?
    echo -e "$(rc_to_colour "$rc")$command\e[0m [$rc] $result"
done << 'COMMANDS'
url_exists http://www.bbc.co.uk
url_http_response_code http://www.bbc.co.uk
url_http_header http://www.bbc.co.uk
COMMANDS

get_user_selection "Abort, retry, fail?" "f" "a" "r" "f"
echo rc=$?

get_user_selection "Abort, retry, fail?" "" "a" "r" "f"
echo rc=$?

get_user_confirmation "Continue?"
echo rc=$?

get_user_confirmation "Continue?" "y"
echo rc=$?

get_user_confirmation "Continue?" "n"
echo rc=$?

